File: app/engine.go
package app

import (
	"context"
	"time"

	"github.com/drydock/drydock/internal/core/domain"
)

// requestTimeout bounds a single engine request initiated from the UI.
const requestTimeout = 15 * time.Second

// ListContainers returns the containers on the given connected host.
func (a *App) ListContainers(hostID string) ([]domain.Container, error) {
	ctx, cancel := a.requestCtx()
	defer cancel()
	eng, err := a.registry.Engine(hostID)
	if err != nil {
		return nil, err
	}
	return eng.ListContainers(ctx)
}

// ListImages returns the images on the given connected host.
func (a *App) ListImages(hostID string) ([]domain.Image, error) {
	ctx, cancel := a.requestCtx()
	defer cancel()
	eng, err := a.registry.Engine(hostID)
	if err != nil {
		return nil, err
	}
	return eng.ListImages(ctx)
}

// ListVolumes returns the volumes on the given connected host.
func (a *App) ListVolumes(hostID string) ([]domain.Volume, error) {
	ctx, cancel := a.requestCtx()
	defer cancel()
	eng, err := a.registry.Engine(hostID)
	if err != nil {
		return nil, err
	}
	return eng.ListVolumes(ctx)
}

// ListNetworks returns the networks on the given connected host.
func (a *App) ListNetworks(hostID string) ([]domain.Network, error) {
	ctx, cancel := a.requestCtx()
	defer cancel()
	eng, err := a.registry.Engine(hostID)
	if err != nil {
		return nil, err
	}
	return eng.ListNetworks(ctx)
}

func (a *App) requestCtx() (context.Context, context.CancelFunc) {
	base := a.ctx
	if base == nil {
		base = context.Background()
	}
	return context.WithTimeout(base, requestTimeout)
}

File: frontend/src/lib/api/engine.ts
// Typed seam over the generated engine bindings. Components import data types
// and calls from here, never from generated paths directly (PROJECT-BOOK §2.8).
import {
  ListContainers,
  ListImages,
  ListNetworks,
  ListVolumes,
} from '../../../wailsjs/go/app/App';
import type { domain } from '../../../wailsjs/go/models';

export type Container = domain.Container;
export type Image = domain.Image;
export type Volume = domain.Volume;
export type Network = domain.Network;

export function listContainers(hostId: string): Promise<Container[]> {
  return ListContainers(hostId);
}

export function listImages(hostId: string): Promise<Image[]> {
  return ListImages(hostId);
}

export function listVolumes(hostId: string): Promise<Volume[]> {
  return ListVolumes(hostId);
}

export function listNetworks(hostId: string): Promise<Network[]> {
  return ListNetworks(hostId);
}
